### DISCUSSION TITLE
RFC: Adaptive blend toevoegen als alternatief voorspelmodel (naast SARIMA→XGBoost)

### DISCUSSION THREAD

**Discussion by Tomeriko96** [cedanl/studentprognose#240] [Ideas]
## Achtergrond

Deze fork wijkt op een aantal punten af van de modellogica in `radboudir/studentprognose`. Na analyse van beide codebases is de conclusie: de afwijkingen zijn bewust en deels beter onderbouwd, maar de upstream heeft ook aanpakken die het waard zijn om aan te bieden.

Het voorstel is **niet** om te mergen (te invasief — de kernformule verschilt fundamenteel), maar om de upstream’s *adaptive blend* toe te voegen als **alternatief naast de huidige SARIMA→XGBoost pipeline**. Instellingen kunnen dan zelf kiezen welke aanpak ze gebruiken, en de vergelijking met upstream wordt reproduceerbaar.

---

### 1. Kernformule voor inschrijvingsprognose (grootste afwijking)

| | Deze fork | Upstream (radboudir) |
|---|---|---|
| **Aanpak** | SARIMA → XGBoost regressor (twee stappen) | Analytische adaptieve blend: `w × signal + (1-w) × prior` |
| **ML in enrollment stap** | XGBoost regressor op SARIMA-geprojecteerde curves | Géén black-box ML — formule met 9 hyperparameters (Optuna-geoptimaliseerd) |

### 2. Ratio-model

- **Fork:** 3-jaars gemiddelde ratio, NF-surplus afgetrokken van NL herkomst
- **Upstream:** Enkel meest recente jaar, NF-surplus proportioneel verdeeld over alle herkomsten

### 3. Prior-selectie

- **Fork:** Geen expliciete prior-selectie — SARIMA draagt dit impliciet
- **Upstream:** R²-gestuurde blend van exponentieel gewogen gemiddelde, naïef vorig jaar, naïeve trend (drempelwaarden 0.30/0.85)

### 4. NF-behandeling

- **Fork:** Aparte XGBoost-run voor NF vs. niet-NF
- **Upstream:** Onderscheidt altijd-NF, vroeger-NF, en hard-gecapped — elk met eigen trainingspool en prior-bron

### 5. Individuele features (XGBoost classifier)

- **Fork heeft meer:** postcode, schoolcodes, adresland
- **Upstream heeft meer:** afstand tot universiteit, NF-aanmeldingsvlaggen, geëngineerte conversieratio’s

### 6. Hardcoded programme-overrides

- **Upstream only:** `B Biology` en `B Psychologie` worden altijd gedwongen naar baseline-only — dit zit niet in de fork

### 7. Individual pipeline

- **Fork:** Volledig actief geïntegreerd
- **Upstream:** Aanwezig maar volledig uitgecommentarieerd (dead code)

**Conclusie:** De afwijkingen zijn niet oppervlakkig — de kern van hoe inschrijvingen worden voorspeld verschilt fundamenteel (ML-keten vs. analytische blend). Tegelijk zijn sommige keuzes in de fork bewust beter onderbouwd (multi-year ratio, robuustheid SARIMA, actieve individual pipeline). Dit is geen achterstand, maar een echte architectuurkeuze die we kunnen uitleggen.

---

### Epic 0 — Architectuurbeslissing (blokker voor alles)

De bestaande registries (`FORECASTER_REGISTRY`, `REGRESSOR_REGISTRY`) werken met numpy arrays. De upstream’s `EnrollmentForecaster` werkt op programmaniveau en heeft per-rij contextuele kolommen nodig (`deadline_week`, `weeks_percentage`). Hij past niet zonder meer in een van de bestaande base classes.

Drie opties:

**Optie A — Strategy-level config flag** *(voorkeur)*
```json
"model_config": {
  "cumulative_pipeline": "sarima_xgboost"  // of "adaptive_blend"
}
```
`CumulativeStrategy` brancht op deze waarde. Twee aparte codepaden, één config-entry. Eenvoudig te begrijpen, maar twee paden om te onderhouden.

**Optie B — Nieuwe `BaseProgrammePredictor` ABC**
Een nieuwe abstractie naast `BaseForecaster`/`BaseRegressor` die op programmaniveau werkt. Cleaner op lange termijn, maar meer initieel werk en een grotere PR.

**Optie C — Force-fit in `BaseRegressor`**
`deadline_week` en `weeks_percentage` meesturen als features in de X-matrix. Hacky, niet aangeraden.

> **❓ Vraag aan de lezer:** welke optie heeft jouw voorkeur? Of zie je een vierde aanpak?

---

### Epic 1 — Preprocessing fundament
*Afhankelijk van: Epic 0*

De adaptive blend heeft twee kolommen nodig die nu niet bestaan in de cumulatieve preprocessing.

**Issue 1.1 — `deadline_week` per rij**
- Startpunt: `strategies/cumulative.py` → `preprocess()`
- Upstream referentie: `src/datasets/cumulative.py` → `Cumulative.deadline_iso_week()` in radboudir/studentprognose
- Wat te bouwen: deadline-datumstrings in config (by_programme, by_exam_type, global fallback) + helper die converteert naar ISO-weeknummer rekening houdend met Studielink refresh offset
- Acceptatiecriterium: `preprocess()` levert een `deadline_week` kolom op per rij; unit tests dekken maandgrens en schrikkeljaar

**Issue 1.2 — `weeks_percentage` per rij**
- Startpunt: `strategies/cumulative.py` → `preprocess()`
- Upstream referentie: `src/datasets/cumulative.py` → `Cumulative._add_weeks_percentage()`
- Wat te bouwen: max-genormaliseerde vulgraad per groep×jaar
- Acceptatiecriterium: waarden liggen in [0, 1]; unit tests dekken lege weken en het 100%-geval

---

### Epic 2 — NF-pool logica uitbreiden
*Kan parallel aan Epic 1*

**Issue 2.1 — Config-schema uitbreiden**
- Startpunt: `config.py` + `configuration.json`
- Breid `numerus_fixus` uit met `historical_nf_years` (dict programme → lijst van jaren) en `hard_capped` vlag
- Acceptatiecriterium: bestaande configs zonder nieuwe keys blijven werken (backwards compatible)

**Issue 2.2 — NF-pool segmentatie in `CumulativeStrategy`**
- Startpunt: `strategies/cumulative.py` → `_predict_students_with_preapplicants()`
- Upstream referentie: `src/core/processor.py` → `DataProcessor.get_training_data()` in radboudir/studentprognose
- Vier segmenten: always-NF pool, used-to-be-NF (eigen history), hard-capped (historisch gemiddelde, geen blend), niet-NF pool
- Extra: Master NL krijgt `year_weight=3` (sterkere recency bias)
- Acceptatiecriterium: unit tests voor elk segment, smoke test passeert

---

### Epic 3 — `AdaptiveBlendPredictor` implementeren
*Afhankelijk van: Epic 0, Epic 1, Epic 2*

Nieuw bestand: `src/studentprognose/models/adaptive_blend.py`
Upstream referentie voor alle sub-issues: `src/prediction_methods/enrollment_forecaster.py` in radboudir/studentprognose

**Issue 3.1 — `compute_prior()`**
- R²-gestuurde blend: R² < 0.30 → 50/50 exp+naïef; R² > 0.85 → 60/40 exp+trend; anders 70/30 exp+naïef
- Acceptatiecriterium: nieuwe opleiding (lege history) geeft `prior = NaN`; unit tests per R²-regime

**Issue 3.2 — `adaptive_weight()`**
- Inputs: `ref_week`, `deadline_week`, `curr_applicants`, historische ratios, `weeks_percentage`
- Componenten: sigmoid tijdvoortgang, fill rate, size penalty, reliability (CV-gebaseerd), fill dampener, post-deadline sigmoid blend
- Acceptatiecriterium: unit tests voor: ruim vóór deadline → w ≈ 0, na deadline → w ≈ 1, klein programma → w gereduceerd

**Issue 3.3 — `predict_programme()`**
- Bouwt op 3.1 + 3.2
- Signal clipping: `clip(signal, last_year ± max(last_year × 0.70, 25))`
- Acceptatiecriterium: integratietest met synthetische data reproduceert een bekend resultaat

**Issue 3.4 — Integratie in `CumulativeStrategy`**
- Implementeer de gekozen architectuur uit Epic 0
- Breid config-validatie uit: `"adaptive_blend"` als geldige waarde
- Acceptatiecriterium: `scripts/test_package.sh` passeert met `cumulative_pipeline: adaptive_blend`

---

### Epic 4 — Optuna hyperparameter tuning *(optioneel, latere fase)*

De upstream optimaliseert de 9 hyperparameters van `adaptive_weight()` via Optuna per instelling. De upstream-defaults zijn bruikbaar als startpunt zonder tuning.

- Voeg `optuna` toe als optionele dependency (`pip install studentprognose[tuning]`)
- Nieuw CLI-subcommando: `studentprognose tune`
- Sla optimale parameters op in configuratiebestand

---

### Epic 5 — Documentatie en tests
*Kan parallel starten zodra Epic 3 begint*

- `docs/methodologie/ensemble.md` — adaptive blend uitleggen naast SARIMA→XGBoost
- `docs/configuratie.md` — nieuwe config keys
- `docs/methodologie/ratio-model.md` — verschil enkelvoudige vs. driejaarlijkse ratio
- Smoke test uitbreiden voor `cumulative_pipeline: adaptive_blend`

---

## Afhankelijkheidsgraph

```
Epic 0
  ├─► Epic 1 (1.1 + 1.2 parallel uitvoerbaar)
  │     └─► Epic 3
  └─► Epic 2 (parallel aan Epic 1)
        └─► Epic 3.4

Epic 4 ── optioneel, na Epic 3
Epic 5 ── start parallel zodra Epic 3 begint
```

## Ruwe inspanning

| Epic | Effort |
|---|---|
| 0 Architectuurbeslissing | 0.5 dag |
| 1 Preprocessing fundament | 3–4 dagen |
| 2 NF-pool logica | 3–4 dagen |
| 3 AdaptiveBlendPredictor | 5–7 dagen |
| 4 Optuna tuning | 3–4 dagen (optioneel) |
| 5 Docs + tests | 2–3 dagen |
| **Totaal (zonder Epic 4)** | **~15–20 dagen** |

---

*Volgende stap: geef je mening over Epic 0. Zodra de architectuurbeslissing vaststaat zetten we de epics om naar losse issues.*

**Comment by Tomeriko96** [comment]
> 🤖 *Deze analyse is gegenereerd door Claude (claude-sonnet-4-6) op basis van een vergelijking van beide codebases. Gebruik als input voor discussie, niet als eindoordeel.*

---

## Mening per afwijkingspunt: fork vs. upstream

### 1. Kernformule — SARIMA→XGBoost vs. analytische blend

**Upstream wint.** Voor een tool die bij meerdere instellingen draait met vaak kleine programma's (< 50 studenten) is interpreteerbaar en overfit-resistent beter dan expressief. XGBoost heeft voldoende trainingsdata per programma nodig — die is er niet altijd. De upstream's adaptieve gewicht lost dit expliciet op via een size penalty. Bovendien: als een voorspelling fout is, kun je bij de analytische blend uitleggen *waarom* (w was hoog want het was laat in het seizoen, signaal was betrouwbaar). Bij XGBoost op SARIMA-curves kun je dat niet.

---

### 2. Ratio-model — 3-jaars gemiddelde vs. enkelvoudig jaar

**Fork wint.** Een enkel uitschietjaar (COVID, eerste jaar van een nieuw programma, beleidswijziging) distorteert de enkelvoudige ratio volledig. Driejaars middeling is standaardpraktijk in ratio-gebaseerde prognoses juist om dit te voorkomen. De upstream is hier te gevoelig voor ruis.

Kanttekening: de NF-cap afhandeling is bij de upstream proportioneel over alle herkomsten — dat is principiëler dan de fork die het surplus alleen van NL herkomst aftrekt.

---

### 3. Prior-selectie — impliciet via SARIMA vs. R²-gestuurd

**Upstream wint.** De R²-gating is expliciet en hanterbaar: sterk historisch patroon → gebruik trend; zwak → val terug op simpeler. Nieuwe opleidingen krijgen automatisch `prior = NaN` → voorspelling puur op signaal. SARIMA als impliciete prior is moeilijker te redeneren en biedt geen graceful fallback voor nieuwe opleidingen.

---

### 4. NF-behandeling — twee groepen vs. vier

**Upstream wint.** In de Nederlandse context wisselen opleidingen regelmatig van NF-status. Een programma dat vorig jaar NF was heeft een fundamenteel ander aanmeldgedrag dan een programma dat al vijf jaar NF is. Die overgangsperiode negeren geeft structureel slechte prognoses. Hard-gecapte programma's op historisch gemiddelde zetten (geen blend) is ook goed gemotiveerd.

---

### 5. Individuele features

**Upstream heeft betere features, fork heeft meer features.** Afstand tot universiteit is een bewezen predictor in de enrollmentliteratuur. NF-aanmeldingsvlaggen vangen competitief aanmeldgedrag (NF als backup vs. eerste keuze) — dat mist de fork volledig. De fork's postcode- en schoolcodes zijn granulairder maar minder theoretisch onderbouwd. Als je er één set mocht kiezen: upstream. Maar ideaal combineer je beide.

---

### 6. Hardcoded programme-overrides

**Geen van beiden** — het is een code smell. Hardgecodeerde uitzonderingen voor specifieke programma's betekenen dat het model ergens structureel faalt en de fix een pleister was. Voor een generalistische tool zijn Radboud-specifieke overrides per definitie niet thuis. De fork heeft ze terecht niet. Maar als deze programma's echt pathologische patronen hebben, lost de fork dat probleem ook nog niet op — het negeert het alleen.

---

### 7. Individual pipeline — actief vs. dead code

**Fork wint.** Individueel-niveau classificatie is methodologisch beter onderbouwd. Dat de upstream het heeft uitgecommentarieerd suggereert dat het in de Radboud-context niet betrouwbaar genoeg was — maar dat is een data- of kalibratieprobleem, geen reden om de aanpak te verwerpen. De fork investeert hier terecht in.

---

## Samenvatting

| Punt | Beter |
|---|---|
| Kernformule | Upstream |
| Ratio-model | Fork |
| Prior-selectie | Upstream |
| NF-behandeling | Upstream |
| Individuele features | Upstream (andere nadruk) |
| Programme-overrides | Fork (niet hebben is beter) |
| Individual pipeline | Fork |

De upstream heeft op de statistische grondslagen vaker het gelijk aan zijn kant. De fork is sterker in robuustheid (ratio) en ambities (individual pipeline). Dat past ook bij de rollen: upstream is een institutioneel prototype, de fork is een generalistische tool die voor meer instellingen moet werken.
